### PR TITLE
Add exec task to agent endpoint

### PR DESCRIPTION
**Pull request by adjackura**

### CODE REVIEW

File: agentendpoint/exec_task.go [hunk 1]
```diff
@@ -71,9 +148,63 @@ func (e *execTask) run(ctx context.Context) error {
 		})
 	}
 
+	stepConfig := e.Task.GetExecStep().GetLinuxExecStepConfig()
+	if goos == "windows" {
+		stepConfig = e.Task.GetExecStep().GetLinuxExecStepConfig()
```
Comment (jaiminsh) [inline agentendpoint/exec_task.go]: This should be GetWindowsExecStepConfig

File: agentendpoint/exec_task.go [hunk 2]
```diff
@@ -16,14 +16,91 @@ package agentendpoint
...
+		return "", errors.New("gcsObject cannot be nil")
+	}
+
+	var reader io.ReadCloser
+	cl, err := storage.NewClient(ctx)
+	if err != nil {
+		return "", fmt.Errorf("error creating gcs client: %v", err)
+	}
+	external.FetchGCSObject(ctx, cl, gcsObject.Object, gcsObject.Bucket, gcsObject.GenerationNumber)
+	defer reader.Close()
```
Comment (wj-chen) [inline agentendpoint/exec_task.go]: reader needs to be assigned, see https://github.com/GoogleCloudPlatform/osconfig/commit/10ef0dbda51d24ea5f2d927d9ea65182f7d0851a, it was dropped after the package common -> package external refactor

File: agentendpoint/exec_task.go [hunk 3]
```diff
@@ -16,14 +16,91 @@ package agentendpoint
...
+	logger.Debugf("Fetched GCS object bucket %s object %s generation number %d", gcsObject.GetBucket(), gcsObject.GetObject(), gcsObject.GetGenerationNumber())
```
Comment (wj-chen) [inline agentendpoint/exec_task.go]: Was thinking about this when preparing for cohort demo the other day...is it ok for us to debug-log the bucket and object names? Are they PII?

Same with some other logging in this file

File: agentendpoint/exec_task.go [hunk 4]
```diff
@@ -16,14 +16,91 @@ package agentendpoint
...
+func executeCommand(path string, exitCodes []int32, args []string, loggingLables map[string]string) (int32, error) {
+	logger.Debugf("Running command %s with args %s", path, args)
+
+	cmd := exec.Command(path, args...)
+	out, err := run(cmd)
+	var exitCode int32
+	if cmd.ProcessState != nil {
+		exitCode = int32(cmd.ProcessState.ExitCode())
+		logger.Infof("Command exit code: %d, stdout:\n%s", exitCode, out)
+	}
```
Comment (wj-chen) [inline agentendpoint/exec_task.go]: I think `cmd.CombinedOutput()` includes stderr from the command too, maybe just update the log message to mention that?

Comment (wj-chen) [reply agentendpoint/exec_task.go]: And will cmd.ProcessState but nil somehow? Will err be guaranteed non-nil in that case (so we go into line 95)?

File: agentendpoint/exec_task.go [hunk 5]
```diff
@@ -16,14 +16,91 @@ package agentendpoint
...
+	localPath := filepath.Join(os.TempDir(), path.Base(gcsObject.GetObject()))
+	if err := external.DownloadStream(reader, "", localPath, 0755); err != nil {
+		return "", fmt.Errorf("error downloading GCS object: %s", err)
+	}
+
+	logger.Debugf("Downloaded to local path %s", localPath)
+	return localPath, nil
+}
+
+func executeCommand(path string, exitCodes []int32, args []string, loggingLables map[string]string) (int32, error) {
```
Comment (wj-chen) [inline agentendpoint/exec_task.go]: exitCodes are not used yet?

File: agentendpoint/exec_task.go [hunk 6]
```diff
@@ -71,9 +148,63 @@ func (e *execTask) run(ctx context.Context) error {
...
+			exitCode, err = executeCommand(localPath, codes, nil, e.LogLabels)
+		}
+	case agentendpointpb.ExecStepConfig_SHELL:
+		if goos == "windows" {
+			exitCode, err = executeCommand(winCmd, codes, []string{"/c", localPath}, e.LogLabels)
+		} else {
+			exitCode, err = executeCommand(sh, codes, []string{localPath}, e.LogLabels)
+		}
+	case agentendpointpb.ExecStepConfig_POWERSHELL:
+		if goos != "windows" {
```
Comment (wj-chen) [inline agentendpoint/exec_task.go]: nit: the other cases check for `if goos == "windows"`

File: agentendpoint/exec_task.go (hunk 4 above)
Comment (adjackura) [reply agentendpoint/exec_task.go]: "ProcessState contains information about an exited process", so I assume we can trust it to be there. 

From the code the only way it wouldn't is if we got an error with the setup of the exec call.

File: agentendpoint/exec_task.go (hunk 3 above)
Comment (adjackura) [reply agentendpoint/exec_task.go]: Logs are all stored in the users project, so logs themselves are PII, so I would say this is fine.

File: agentendpoint/exec_task.go (hunk 5 above)
Comment (adjackura) [reply agentendpoint/exec_task.go]: Dropped them from the agent code as this should be handled in the service

**Pull request merged**